### ISSUE TITLE
ETIMEDOUT in sendNotification.Can it add a proxy option?

### ISSUE THREAD

**Issue by yyh1102** [web-push-libs/web-push#280]
I've already used webpush.sendNotification and it automatically request the Google server address.But I received so many ETIMEDOUT. 
![image](https://cloud.githubusercontent.com/assets/10328169/26300413/9962a8fa-3f10-11e7-9631-a6d6524309a9.png)

For the server in China, the network is unluckily blocked by gov, so we have to use proxy server to link websites abroad. 

**Hope the web-push can add proxy options.**

**Comment by marco-c** [comment]
What options would you need?

**Comment by yyh1102** [comment]
Like
```
const options={
  ...,
  proxy:"http://xxx.xxx.xxx"  // my proxy server
}
webpush.sendNotification(
  pushSubscription,
  payload,
  options
);
```
So that the web-push can request the google url through my proxy server.

**Comment by marco-c** [comment]
I'd happily accept a pull request for this.

You can also implement the same without using a standard proxy. Just change the endpoint URL to be the URL of your custom proxy (e.g. YOUR_PROXY_URL?endpoint=ACTUAL_ENDPOINT_URL) and make the proxy send the request to the endpoint URL.

**Comment by yyh1102** [comment]
Oh,that's a good idea~ But actually the proxy server is not mine, and i can do nothing on it to build such custom api.
A pr can be ok~
Thanks for replying.

**Comment by yyh1102** [comment]
Hey due. I've forked your repo and add proxy api. But after serious consideration, i think it's not necessary to pull request. There are three reasons:
- The demand of proxy is just for Chinese developers, so this api will be very duplicate for other people.
- The original code can not be run in my local environment for some reasons, so I can't do automatic test.
- For convenience I only change https to http to request the Push Server using proxy, and it works in my project. 

So it's better to make it an independent repo, and this is the [link](https://github.com/yyh1102/web-push-China)

**Comment by MartijnDwars** [comment]
Note that it's not only Chinese developers that are affected by this problem. If you are behind a corporate proxy, which is common in big organisations, you encounter the same issue. Having the ability to set a proxy, or better, respecting the HTTP(S)_PROXY environment variables, would benefit these users. That said, #282 provides exactly this functionality, so I think the focus should be on integrating that PR.

**Comment by yyh1102** [comment]
Oh. I have not noticed that PR. It's really good if the official lib implements that API. Thanks for contributors!!

**Comment by Peanoquio** [comment]
Hi there,
Pardon me for posting this late since this topic has already been up for a few months.
Just to confirm, does this mean that the web push library cannot work behind a corporate proxy??
Thank you in advance.

**Comment by marco-c** [comment]
I think so, until https://github.com/web-push-libs/web-push/pull/282 is finished up and merged.

**Comment by Peanoquio** [comment]
Thank for the update regarding this.
Have a great week ahead!

**Comment by marco-c** [comment]
@anasameur if you want, you can pick up the already existing PR (https://github.com/web-push-libs/web-push/pull/282) and finish it.

**Comment by marco-c** [comment]
Fixed by #319.
